### DISCUSSION TITLE
How to call async methods on tab/window close?

### DISCUSSION THREAD

**Discussion by Wizzel1** [wxt-dev/wxt#915] [Q&A]
So I am trying to set up listeners for `browser.tabs.onRemoved` and  `browser.windows.onRemoved` in my background script. 

These listeners need to be async because I am trying to perform a database operation and set local storage properties. 

Something like this:

```ts
browser.windows.onRemoved.addListener(async (windowId) => {
  const uploaderTab = await extensionStorage.getItem('uploaderTab');
  if (windowId === uploaderTab?.windowId) {
    console.log('Closed the uploader window', windowId); //this logs every time
    await resetTasks(); // sometimes has no effect.
    await extensionStorage.removeItem('uploaderTabId');
    await extensionStorage.removeItem('autoRunUploader');
   console.log("always gets logged");
  }
});
```

but that isn't stable. Sometimes it works, sometimes it doesn't. The `resetTasks()` methods runs a supabase query.

Am I doing something wrong or is there a more reliable way to do that? 

I also tried everything in this SO post with the same result:
https://stackoverflow.com/a/70802055/10847224

**Comment by aklinker1** [comment]
Code you shared looks fine. Consider adding a mutex to make sure your logic isn't being ran in multiple times in parallel, but closing windows should happen rare enough this isn't contributing to your problem.

```ts
import { Mutex } from 'async-mutex';

const mutex = new Mutex();
browser.windows.onRemoved.addListener((windowId) => mutex.runExclusive(() => {
  // your logic here
}));
```

Otherwise, it seems to me like `resetTasks` could be problematic, but I don't know what it looks like. If `"Closed the uploader window"` and `"always gets logged"` are always logged, then you are calling `resetTasks` correctly, which is why I think that function is broken.

**Reply by aklinker1** [reply]
I saw on discord you're seeing this error message:

```
A listener indicated an asynchronous response by returning true, but the message channel closed before a response was received
```

Are you trying to send a message in `resetTasks`?

**Reply by Wizzel1** [reply]
> I saw on discord you're seeing this error message:
> 
> ```
> A listener indicated an asynchronous response by returning true, but the message channel closed before a response was received
> ```
> 
> Are you trying to send a message in `resetTasks`?

No, for some reason that message didn't came up again. 

Regarding the initial post: You are right, apparently this didn't have to do anything with the callback but rather that the supabase client didn't have a valid session. Its really annoying that supabase doesn't throw an error in that case... 

I will do some further testing and let you know if I encounter the same problem again. Sorry for the dumb questions...

**Reply by Wizzel1** [reply]
So my method works now but there is something else I noticed. The `onRemove` listeners don't fire when the entire browser is closed (or the last tab of a browser). I guess this is expected but do you know of any trick that could help me with this?

**Reply by aklinker1** [reply]
There are on suspend and startup listeners you can add to clear your storage

- https://developer.chrome.com/docs/extensions/reference/api/runtime#event-onStartup
- https://developer.chrome.com/docs/extensions/reference/api/runtime#event-onSuspend
